### ISSUE TITLE
[Integration] Generator Runtime Prediction 운영 연결 및 기본 Model Artifact Bootstrap

### ISSUE THREAD

**Issue by smmini** [Biz-CollabCraft/ontology_dashboard#131]
## 목적

PR #127 이후 Generator Runtime Prediction을 실제 운영·시연 가능한 전체 흐름으로 연결하고, 재현 가능한 기본 Model Artifact Bootstrap을 제공한다.

이 이슈는 다음 전체 통합을 추적한다.

```text
gen_data / Runtime Overlay
→ Generator enqueue
→ Preprocessing / Runtime Feature
→ Model inference
→ Prediction Result Batch
→ Backend Prediction Inbox
→ Threshold / Decision Policy
→ Product Result Artifact
→ Evidence / Report
→ Dashboard
```

Generator는 raw score 생성과 Outbox 전달까지 담당한다. Threshold, 최종 이상 판정, Product Result Artifact, Evidence 및 Report는 Backend가 담당한다. Frontend와 Closed-loop는 Generator의 raw Prediction Result Batch를 직접 소비하지 않는다.

## 현재 상태

- PR #127은 Generator Runtime Prediction, Active Model Set, Model Artifact 검증, checkpoint 및 Outbox 경계를 구현 중이다.
- Runtime은 기본적으로 비활성화되어 있다.
- Backend Prediction Inbox와 전체 승격 경로는 아직 구현되지 않았다.
- 저장소에 재현 가능한 기본 Model Artifact Bootstrap은 아직 없다.

## 선행조건 — PR #127

다음 항목이 PR #127에서 완료되고 병합된 이후 이 이슈의 통합 작업을 시작한다.

- [ ] 외부 Current 계약을 `contracts/schemas/prediction-result-batch.schema.json`의 `prediction-result-batch-v1`으로 통일
- [ ] Generator 내부 K-V 결과를 외부 `results[]` Batch로 변환
- [ ] required 모델 inference 실패를 fail-closed 처리
- [ ] inference 실패가 로그만 남고 결과에서 누락되는 동작 제거
- [ ] 배열형 Contract Example·Vector·Runtime 테스트 통과
- [ ] 기존 K-V Schema의 외부 Current 지위 제거 또는 Internal staging으로 강등

PR #127에 위 작업이 반영되기 전에는 “배열형 Prediction Result Batch 및 명시적 inference 실패 처리가 적용됐다”고 기록하지 않는다.

## A. 기본 Model Artifact Bootstrap

- [ ] 초기 promoted base model 1개 결정
- [ ] 현재 등록된 모델 ID와 canonical naming 확인
- [ ] 고정 Dataset, Feature Dataset, Label Schema 및 Training Config 결정
- [ ] 고정 random seed와 split strategy 기록
- [ ] 기본 모델 학습 실행
- [ ] 불변 Model Artifact 6개 파일 발행
- [ ] 공식 Manifest Schema 및 모든 SHA-256 검증
- [ ] `active-model-set.json` 원자적 발행
- [ ] 재현 가능한 bootstrap 명령 또는 배포 entrypoint 제공
- [ ] Artifact 배포 위치와 volume 권한 정책 기록
- [ ] 재실행 시 동일 버전 덮어쓰기 차단

필수 Artifact:

```text
manifest.json
model.joblib
feature_schema.json
label_schema.json
history_requirement.json
metrics.json
```

초기에는 단일 promoted 모델을 사용한다. LightGBM, XGBoost, Random Forest 중 실제 등록·배포 의존성이 확인된 모델을 선택한다. CatBoost는 별도 등록과 계약·테스트가 추가되기 전까지 기본 Bootstrap 대상에 포함하지 않는다.

## B. Generator 송신 설정 정합화

- [ ] 환경변수를 `GENERATOR_PREDICTION_RESULT_URL`로 통일
- [ ] canonical Backend endpoint를 `POST /internal/prediction-results`로 통일
- [ ] 기존 `GENERATOR_PREDICTION_RECEIVER_URL`의 제거 또는 deprecation 정책 결정
- [ ] 배포 환경별 Backend base URL 구성
- [ ] timeout, retry, backoff 및 dead-letter 설정 검증
- [ ] Generator → Backend HTTP Contract Test 추가

## C. Backend Prediction Inbox

- [ ] `POST /internal/prediction-results` 구현
- [ ] `prediction-result-batch-v1` 공식 Schema 검증
- [ ] `batch_id` 및 item `event_id` 멱등 저장
- [ ] 동일 ID·동일 payload 재전송 안전 수용
- [ ] 동일 ID·다른 payload conflict 차단
- [ ] item 및 Batch checksum 검증
- [ ] `output_status`별 수신 정책 구현
- [ ] 원본 upstream Batch append-only 보존
- [ ] Inbox 저장과 Product Result 승격 책임 분리
- [ ] Backend가 Generator 내부 파일 경로를 직접 탐색하지 않도록 차단

## D. Runtime Overlay 자동 연결

- [ ] `runtime_overlay.observations.available` consumer 또는 watcher 구현
- [ ] 완료 파일만 소비하도록 atomic publish/checksum 검증
- [ ] source cursor 저장
- [ ] 중복 enqueue 차단
- [ ] 누락 tick 및 순서 역전 감지
- [ ] Generator Pipeline 자동 enqueue
- [ ] 수동 enqueue와 자동 enqueue의 운영 경계 구분
- [ ] 재시작 후 미처리 입력 재개

## E. Maintenance 및 Overlay Lineage

Prediction Result item에서 다음 lineage를 끝까지 보존한다.

- [ ] `simulation_session_id`
- [ ] `overlay_branch_id`
- [ ] `history_segment_id`
- [ ] `maintenance_event_id`
- [ ] `maintenance_action_id`
- [ ] `state_version`
- [ ] `source_kind`
- [ ] `source_ref.uri`
- [ ] `source_ref.sha256`

추가 검증:

- [ ] `maintenance_replay`에서 필수 lineage 누락 차단
- [ ] 정비 전·후 Feature history segment 분리
- [ ] 정비 전 Result 불변 보존
- [ ] 정비 후 Observation에 대한 별도 Prediction Result 생성

## F. Backend 승격 경계

- [ ] 모델별 raw score 검증
- [ ] Threshold Policy 적용
- [ ] Model Set 및 필수 모델 결과 누락 검증
- [ ] 설비별 최종 Decision 생성
- [ ] Product Result Artifact 생성
- [ ] Evidence 생성
- [ ] Report 생성
- [ ] Dashboard ViewModel 제공
- [ ] Frontend와 Closed-loop의 raw Batch 직접 소비 차단
- [ ] 실패·warming-up·history-insufficient 상태의 Product 표현 결정

## G. 기존 Backend Direct Inference 전환

- [ ] 기존 Backend direct inference 경로 목록 조사
- [ ] 기존 경로를 즉시 제거하지 않고 shadow mode 유지
- [ ] Generator와 Backend score parity 검증
- [ ] Feature Schema 및 row alignment parity 검증
- [ ] Model Artifact 및 lineage parity 검증
- [ ] rollback 경로 검증
- [ ] Generator canonical Runtime 활성화
- [ ] 기존 Backend direct inference 비활성화
- [ ] 안정화 이후 레거시 코드 제거

전환 순서:

```text
Shadow 수신
→ score·feature·lineage parity
→ Backend 승격 E2E
→ rollback 검증
→ Generator canonical 활성화
→ 기존 Backend inference 비활성화
```

## H. Runtime 활성화 및 E2E

- [ ] 기본값 `GENERATOR_RUNTIME_PREDICTION_ENABLED=false` 유지
- [ ] Backend consumer 준비 전 활성화 금지
- [ ] 통합 테스트 환경에서만 우선 활성화
- [ ] Generator → Backend Contract Test
- [ ] Generator → Backend Inbox → Product Result 통합 테스트
- [ ] Evidence·Report 생성 검증
- [ ] Dashboard 표시 검증
- [ ] Outbox retry/backoff/dead-letter 검증
- [ ] 프로세스 재시작 및 checkpoint 재개 검증
- [ ] 중복·부분 실패·checksum conflict 검증
- [ ] 운영 활성화 승인과 rollback 절차 기록

## 범위 밖

- Threshold 또는 Product Result 판단을 Generator로 이동
- Frontend가 raw Prediction Result Batch를 직접 해석
- 미등록 모델을 기본 모델로 자동 선택
- `model.joblib`만 단독으로 Git에 커밋
- Runtime 준비 전 feature flag 기본값을 `true`로 변경
- 검증 없이 기존 Backend direct inference 선삭제

## 완료 조건

- [ ] 재현 가능한 단일 기본 Model Artifact와 Active Model Set이 발행된다.
- [ ] Generator가 Runtime Overlay 입력을 자동 enqueue한다.
- [ ] Backend Prediction Inbox가 배열형 정본을 검증하고 멱등 저장한다.
- [ ] Generator 송신 설정과 Backend endpoint가 하나의 계약으로 통일된다.
- [ ] Maintenance/Overlay lineage가 Prediction부터 Product Result까지 보존된다.
- [ ] Backend가 raw score를 Threshold·Product Result·Evidence·Report로 승격한다.
- [ ] 기존 Backend direct inference가 shadow 검증과 rollback 준비 후 전환된다.
- [ ] Generator → Backend → Dashboard E2E가 통과한다.
- [ ] Runtime 기본 비활성화 및 환경별 활성화 정책이 문서화된다.

<details>
<summary>본문 변경 기록</summary>

### 2026-08-27 — PR #127 후속 운영 통합 범위 정합화

- PR #127에서 아직 완료되지 않은 배열형 계약과 inference fail-closed를 선행조건으로 정정
- 초기 기본 모델을 등록된 단일 promoted 모델로 제한
- 미등록 CatBoost 자동 Bootstrap 항목 제거
- Backend endpoint를 `/internal/prediction-results` 기준으로 정정
- Runtime Overlay enqueue, Maintenance lineage, Backend Inbox, 승격, shadow 전환 및 E2E 체크포인트 추가
- Runtime 기본 비활성화와 Backend 소비 준비 전 활성화 금지 명시

</details>

**Comment by enjoylonelines** [comment]
## Backend Prediction Inbox 작업 제안

Issue #131의 통합 흐름 중 **Backend Prediction Inbox / Product Result 승격 Gate** 영역은 제가 맡아서 진행하겠습니다.

### 제안 배경

현재 PR #127은 Generator Runtime Prediction, `prediction-result-batch-v1` 계약, Outbox 전달 경계를 준비하는 upstream producer 작업입니다. 다만 Issue #131에도 정리되어 있듯이, Backend가 해당 Batch를 수신·검증·보존하고 Product Result/Evidence로 승격하는 경로는 아직 구현되지 않았습니다.

따라서 다음 단계에서는 Generator 결과를 바로 UI나 Closed-loop가 소비하지 않고, Backend가 명확한 Inbox gate를 통해 검증한 뒤 Product Result로 승격하는 구조를 먼저 닫는 것이 맞다고 봅니다.

### 제가 맡을 범위

- `POST /internal/prediction-results` Backend 수신부 구현
- `prediction-result-batch-v1` payload 검증
- `batch_id`, item `event_id`, `payload_sha256` 기반 멱등성 처리
- 동일 payload 재전송 허용
- 동일 ID / 다른 payload conflict 차단
- 원본 upstream Batch append-only 보존
- Inbox 저장과 Product Result 승격 책임 분리
- `output_status`별 수신 정책 정리
- Backend가 Generator 내부 파일 경로나 구현을 직접 탐색하지 않도록 차단
- Product Result 승격 전 raw score가 Frontend / Closed-loop에 노출되지 않도록 경계 유지

### 명확히 범위 밖

- Generator가 Threshold 또는 최종 이상 판정을 수행하도록 변경
- Frontend가 raw Prediction Result Batch를 직접 해석
- Backend direct inference를 즉시 제거
- Product Result / Evidence core 의미 변경
- Runtime feature flag 기본값을 바로 활성화

### 완료 기준

- Backend가 PR #127의 `contracts/schemas/prediction-result-batch.schema.json` 계약을 기준으로 Batch를 검증한다.
- 같은 Batch 재전송은 중복 Product Result를 만들지 않는다.
- 같은 ID에 다른 payload가 들어오면 conflict로 차단한다.
- `history_insufficient`, `warming_up`, `failed_*` 상태는 정상 Product Result로 보정되지 않는다.
- Product Result 승격 전까지 raw score는 Product API/UI/Closed-loop trigger로 노출되지 않는다.
- 기존 Product Result/Evidence 테스트와 신규 Inbox contract/idempotency 테스트가 함께 통과한다.

### 진행 방식

PR #127이 외부 wire contract를 정리한 뒤, 저는 Backend Inbox를 별도 작업 단위로 시작하겠습니다.  
초기 목표는 “수신 + 검증 + 멱등 저장 + 승격 전 경계”를 먼저 닫고, Product Result Artifact 생성 및 Dashboard 연결은 같은 stack의 다음 단계로 이어가겠습니다.

**Comment by KOR-GANG** [comment]
@smmini PR #127 머지 이후 PR #132와 연결하기 위해 Generator ingress/pass-through 경계에서 추가 확인이 필요한 내용을 공유드립니다.

현재 PR #132는 `runtime_overlay.observations.available`을 검증하고 누적 history를 immutable snapshot으로 고정한 뒤, 다음 source context를 `/internal/runtime-pipeline/enqueue` payload에 포함하도록 구현되어 있습니다.

- `source_kind=maintenance_replay_overlay`
- `simulation_session_id`
- `overlay_branch_id`
- `history_segment_id`
- `maintenance_event_id`
- `maintenance_action_id`
- `state_version`

다만 현재 main의 Generator ingress는 해당 값을 수신·보존할 공식 경로가 아직 없습니다.

- `EnqueueRequest`에 `source_kind`와 구조화된 lineage가 없음
- 알 수 없는 필드가 silent ignore됨
- `PipelineQueueItem` 및 SQLite Queue에 source context 저장 필드가 없음
- claim/retry/recovery/checkpoint에서 lineage를 보존할 수 없음
- Stage와 delivery fallback 경로가 `live_sensor`와 빈 lineage를 사용
- Feature/History version 누락 시 `v1`/`v1.0`을 생성하는 fallback이 남아 있음

Generator 쪽 후속 범위는 다음과 같이 보면 좋겠습니다.

1. EnqueueRequest에 canonical `source_kind`와 구조화된 lineage 추가 및 extra field fail-closed
2. Queue 영속화와 claim/retry/recovery/checkpoint에서 동일 source context 보존
3. Stage 및 외부 `prediction-result-batch-v1`까지 입력 lineage를 변환 없이 전달
4. Feature/History version의 임의 `v1`/`v1.0` fallback 제거
5. Overlay enqueue → Queue → external Prediction Result Batch lineage 회귀 테스트 추가

PR #132에서는 Generator 내부 구현을 수정하지 않고 Runtime Overlay event 소비, immutable snapshot 및 공식 HTTP enqueue payload 구성까지만 담당하겠습니다. 위 Generator pass-through가 보완되었다는 전제에서 PR #132를 최신 main에 맞추고 실제 lineage round-trip을 검증하겠습니다.

Backend Prediction Inbox와 Product Result/Evidence 승격은 기존 #131 담당 범위를 그대로 따릅니다.

Refs #131
Related to #117

**Comment by KOR-GANG** [comment]
Issue #99 종료를 위한 Runtime Integration 인계 상태를 현행화합니다.

## 이후 완료된 기반

- PR #127: Generator Runtime Prediction Pipeline과 `prediction-result-batch-v1`
- PR #132: 정비 후 Overlay immutable snapshot → Generator enqueue
- PR #134: Runtime source context/provenance 보존
- PR #135: Backend receive-only Prediction Inbox
- PR #139: Overlay identity/checksum/branch lineage 및 PostgreSQL 동시성 방어
- PR #140: Inbox Batch → Backend Threshold/정책 → Product Result/Evidence 승격

## 이 이슈에서 계속 추적할 미완료 범위

- 일반 Canonical/Live Observation의 Generator Runtime Pipeline enqueue
- 기존 Backend direct inference의 shadow parity 검증과 비활성화/제거
- 재현 가능한 promoted Model Artifact/Active Model Set bootstrap
- Generator runtime 운영 활성화와 rollback 절차
- PostgreSQL/RLS/shared mount/지속 worker/service token 기반 배포형 E2E
- Generator → Backend → Product Result/Evidence → Dashboard의 장애·재시작·멱등 회귀

Product API의 상세 Closed-loop lineage, Runtime Status 및 정비 전·후 Result 조회는 #153으로 분리했습니다.

따라서 #99를 닫더라도 위 Runtime 전환·배포 작업은 이 이슈에서 계속 추적합니다.
